test/e2e/gateway_api_test.go

```diff
@@ -0,0 +1,266 @@
+	t.Helper()
+
+	if err := assertSubscription(t, openshiftOperatorsNamespace, expectedSubscriptionName); err != nil {
+		t.Fatalf("failed to find expected Subscription %s: %v", expectedSubscriptionName, err)
+	}
+	if err := assertCatalogSource(t, expectedCatalogSourceNamespace, expectedCatalogSourceName); err != nil {
+		t.Fatalf("failed to find expected CatalogSource %s", expectedCatalogSourceName)
+	}
+	if err := assertOSSMOperator(t); err != nil {
+		t.Fatal("failed to find expected Istio operator")
+	}
+	if err := assertIstiodProxy(t); err != nil {
+		t.Fatal("failed to find expected Istiod proxy")
+	}
+	// TODO - In OSSM 3.x the configuration object to check will be different.
+	if err := assertSMCP(t); err != nil {
+		t.Fatalf("failed to find expected SMCP: %v", err)
+	}
+}
+
+// testGatewayAPIObjects tests that Gateway API objects can be created successfully.
+func testGatewayAPIObjects(t *testing.T) {
+	t.Helper()
+
+	// Create a test namespace that cleans itself up and sets up its own service account and role binding.
+	ns := createNamespace(t, names.SimpleNameGenerator.GenerateName("test-e2e-gwapi-"))
+
+	// Update the ingress-operator cluster role with cluster-admin privileges.
+	// TODO - Should not be needed in OSSM 2.5, but for now it still is. See https://issues.redhat.com/browse/OSSM-3508.
+	if err := updateIngressOperatorRole(t); err != nil {
+		t.Fatalf("failed to update ingress operator role: %v", err)
+	}
+
+	// Validate that Gateway API objects can be created.
+	if err := ensureGatewayObjectCreation(ns); err != nil {
+		t.Fatalf("failed to create one or more gateway object/s: %v", err)
+	}
+
+	// Wait for the Gateway API objects to reach a successful status.
+	errors := ensureGatewayObjectSuccess(t, ns)
+	if len(errors) > 0 {
+		t.Errorf("failed to observe successful status of one or more gateway object/s: %v", strings.Join(errors, ","))
+	} else {
+		t.Log("gateway class, gateway, and http route created successfully")
+	}
+}
+
+func ensureCRDs(t *testing.T) {
+	t.Helper()
+	for _, crdName := range crdNames {
+		crdVersion, err := assertCrdExists(t, crdName)
+		if err != nil {
+			t.Fatalf("failed to find crd %s: %v", crdName, err)
+		}
+		t.Logf("found crd %s at version %s", crdName, crdVersion)
+	}
+}
+
+// ensureGatewayObjectCreation tests that gateway class, gateway, and http route objects can be created.
+func ensureGatewayObjectCreation(ns *corev1.Namespace) error {
+	var domain string
+
+	gatewayClass, err := createGatewayClass(gatewayclass.OpenShiftDefaultGatewayClassName, gatewayclass.OpenShiftGatewayClassControllerName)
+	if err != nil {
+		return fmt.Errorf("feature gate was enabled, but gateway class object could not be created: %v", err)
+	}
+	// We don't need to delete the gateway class so there is no defer function for it.
+
+	// Use the dnsConfig base domain set up in TestMain.
+	if dnsConfig.Spec.BaseDomain != "" {
+		domain = "gws." + dnsConfig.Spec.BaseDomain
+	} else {
+		domain = ".example.com"
+	}
```

Comment:
*nit* I think it's safe to assume that `dnsConfig.Spec.BaseDomain` exists like we do in other tests. The odd thing is, if for some reason it doesn't exist, we set the domain to `.example.com`, then the rest of the test will fail because we can't create DNS records for that domain.

Reply:
Fixed.